frontends/common/constantFolding.cpp

```diff
@@ -725,13 +725,22 @@ const IR::Node *DoConstantFolding::postorder(IR::Concat *e) {
     auto eright = getConstant(e->right);
     if (eleft == nullptr || eright == nullptr) return e;
 
-    auto left = eleft->to<IR::Constant>();
-    if (left == nullptr) {
-        // Could be a SerEnum
-        return e;
+    const auto *lstr = eleft->to<IR::StringLiteral>();
+    const auto *rstr = eright->to<IR::StringLiteral>();
+
+    const auto *left = eleft->to<IR::Constant>();
+    const auto *right = eright->to<IR::Constant>();
+
+    // handle string concatenations
+    if (lstr || rstr) {
+        if (!(lstr && rstr)) {
+            error(ErrorType::ERR_INVALID, "%1%: cannot concatenate string and non-string", e);
+            return e;
```

Comment:
Hmm, I've added this error because this pass runs before type checking. But I guess I can instead just not fold these and let type checker do the rest. That would be cleaner. I'll test the variable too.

Reply:
Done. There is already a test checking that prohibits `string` variables: [`testdata/p4_16_errors/string-e1.p4`](https://github.com/p4lang/p4c/blob/main/testdata/p4_16_errors/string-e1.p4), so I don't think we need another one that also does concats.